source/iterator/snapshot.go

```diff
@@ -0,0 +1,231 @@
+		return nil, fmt.Errorf("vtgateconn dial: %w", err)
+	}
+
+	target := strings.Join([]string{params.Keyspace, params.TabletType}, "@")
+	session := conn.Session(target, &query.ExecuteOptions{
+		IncludedFields: query.ExecuteOptions_ALL,
+	})
+
+	snapshot := &Snapshot{
+		conn:           conn,
+		session:        session,
+		records:        make(chan sdk.Record, defaultRecordsBufferSize),
+		table:          params.Table,
+		keyColumn:      params.KeyColumn,
+		orderingColumn: params.OrderingColumn,
+		columns:        params.Columns,
+		batchSize:      params.BatchSize,
+	}
+
+	if params.Position != nil {
+		snapshot.position = params.Position
+	}
+
+	if err = snapshot.loadRecords(ctx); err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	return snapshot, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (s *Snapshot) HasNext(ctx context.Context) (bool, error) {
+	if len(s.records) == 0 {
+		if err := s.loadRecords(ctx); err != nil {
+			return false, fmt.Errorf("load records: %w", err)
+		}
+	}
+
+	return len(s.records) > 0, nil
+}
+
+// Next returns the next record.
+func (s *Snapshot) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+
+	case record := <-s.records:
+		return record, nil
+	}
+}
+
+// Stop closes the underlying db connection.
+func (s *Snapshot) Stop(ctx context.Context) error {
+	if s.conn != nil {
+		s.conn.Close()
+	}
+
+	return nil
+}
+
+// loadRecords selects a batch of rows from a database, based on the Snapshot's
+// table, columns, orderingColumn, batchSize and the current position,
+// and converts them to the sdk.Record.
+func (s *Snapshot) loadRecords(ctx context.Context) error {
+	selectBuilder := sqlbuilder.NewSelectBuilder()
+
+	if len(s.columns) > 0 {
+		selectBuilder.Select(s.columns...)
+	} else {
+		selectBuilder.Select("*")
+	}
+
+	selectBuilder.From(s.table)
+
+	if s.position != nil {
+		selectBuilder.Where(
+			selectBuilder.GreaterThan(s.orderingColumn, s.position.LastProcessedElementValue),
+		)
+	}
+
+	selectBuilder.
+		OrderBy(s.orderingColumn).
+		Limit(s.batchSize)
+
+	sql, args := selectBuilder.BuildWithFlavor(sqlbuilder.MySQL)
+
+	query, err := sqlbuilder.MySQL.Interpolate(sql, args)
+	if err != nil {
+		return fmt.Errorf("interpolate arguments to SQL: %w", err)
+	}
+
+	resultStream, err := s.session.StreamExecute(ctx, query, nil)
+	if err != nil {
+		return fmt.Errorf("stream execute select query: %w", err)
+	}
+
+	if err := s.processStreamResults(ctx, resultStream); err != nil {
+		return fmt.Errorf("process stream results: %w", err)
+	}
+
+	return nil
+}
+
+// processStreamResults receives a result from the sqltypes.ResultStream,
+// converts its rows to the sdk.Record and sends them to the i.records channel.
+func (s *Snapshot) processStreamResults(ctx context.Context, resultStream sqltypes.ResultStream) error {
+	for {
+		result, err := resultStream.Recv()
```

Comment:
it is can be question about handling errors. Could we check what error and do we need to do retry

Reply:
Here can be two types of errors, the first is `io.EOF` which is fine, and the second type is something wrong with the stream and we can lose data, so I just return an error here